modules/ppcp-wc-gateway/src/Processor/OrderProcessor.php

```diff
@@ -212,84 +212,96 @@ public function __construct(
 	 * @throws Exception If processing fails.
 	 */
 	public function process( WC_Order $wc_order ): void {
-		$order = $this->session_handler->order();
-		if ( ! $order ) {
-			// phpcs:ignore WordPress.Security.NonceVerification
-			$order_id = $wc_order->get_meta( PayPalGateway::ORDER_ID_META_KEY ) ?: wc_clean( wp_unslash( $_POST['paypal_order_id'] ?? '' ) );
-			if ( is_string( $order_id ) && $order_id ) {
-				try {
-					$order = $this->order_endpoint->order( $order_id );
-				} catch ( RuntimeException $exception ) {
-					throw new Exception( __( 'Could not retrieve PayPal order.', 'woocommerce-paypal-payments' ) );
-				}
-			} else {
-				$is_paypal_return = isset( $_GET['wc-ajax'] ) && wc_clean( wp_unslash( $_GET['wc-ajax'] ) ) === 'ppc-return-url'; // phpcs:ignore WordPress.Security.NonceVerification.Recommended
-
-				if ( $is_paypal_return ) {
-					$this->logger->warning(
-						sprintf(
-							'No PayPal order ID found for WooCommerce order #%d.',
-							$wc_order->get_id()
+		if ( ! $this->verify_order_can_be_processed( $wc_order ) ) {
+			return;
+		}
+
+		$wc_order->update_meta_data( '_ppcp_processing', 'yes' );
```

Comment:
Suggestion: I'd consider adding a timestamp-based lock expiration (to prevent a situation when PHP process crashes `_ppcp_processing` flag might remain set forever, blocking all future attempts):
```
  // When setting lock:
  $wc_order->update_meta_data( '_ppcp_processing', time() );
```

```
  // When checking lock:
  $processing_lock = (int) $wc_order->get_meta( '_ppcp_processing', true );
  if ( $processing_lock && ( time() - $processing_lock ) < 60 ) {
      $this->logger->warning( /* lock active */ );
      return false;
  }
```